glean-core/rlb/src/configuration.rs

```diff
@@ -19,4 +24,8 @@ pub struct Configuration {
     pub delay_ping_lifetime_io: bool,
     /// The release channel the application is on, if known.
     pub channel: Option<String>,
+    /// The server pings are sent to.
+    pub server_endpoint: Option<String>,
+    /// The client implementation to use for uploading pings.
+    pub http_client: Option<Box<CustomUploaderType>>,
```

Comment:
@badboy @brizental I'm trying to extend the `Configuration` to allow for custom uploader to be passed in. What I'd love to achieve:

- The developers should be able to pass in an instance of any object implementing the `net::PingUploader` trait;
- If no custom uploader is provided, we'll provide a default simple HTTPS uploader (coming as a separate PR).

The problem with the current implementation is that, by using generics, `glean.initialize` requires us to tell the type for `Configuration` (see my other comment)

Reply:
You can use a trait object here, given that it's boxed anyway.

```rust
http_client: Option<Box<dyn net::PingUploader>
```

Reply:
That was my first attempt: unfortunately I got shouted at with this:

```
`(dyn net::PingUploader + 'static)` doesn't implement `std::fmt::Debug`

`(dyn net::PingUploader + 'static)` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`

help: the trait `std::fmt::Debug` is not implemented for `(dyn net::PingUploader + 'static)`
note: required because of the requirements on the impl of `std::fmt::Debug` for `std::boxed::Box<(dyn net::PingUploader + 'static)>`
note: required because of the requirements on the impl of `std::fmt::Debug` for `std::option::Option<std::boxed::Box<(dyn net::PingUploader + 'static)>>`
...
```

Which got me wandering down the path of figuring out if there's a way to exclude specific fields from the `Debug` and `Clone` directives, with no luck.

Reply:
There's no automatic way, but we can implement `Debug` manually then. I'll take a closer look after the glean weekly on what need to do here.

Reply:
We can also make `trait PingUploader: Debug`, thus requiring all ping uploaders to implement it, then the derive still works.

Reply:
This way there's no more generics in `Configuration` and we can store a boxed uploader. It's a trait object and thus dynamic dispatch, but I don't think this is any sort of issue.

Reply:
> 
> 
> This way there's no more generics in `Configuration` and we can store a boxed uploader. It's a trait object and thus dynamic dispatch, but I don't think this is any sort of issue.

@badboy what do you mean, here?